### DISCUSSION TITLE
Problem related to fd

### DISCUSSION THREAD

**Discussion by seo-rii** [ANLAB-KAIST/KENSv3#224] [Project Q&A 2. Transport Layer (TCP)]
In some case, fd assigned by createFileDescriptor makes error while calling removeFileDescriptor. (e.g. TestTransfer_Accept_Send_Symmetric). Some case does not make this error.

It creates  `Assertion failed: (nsIter != appIter->second.fdToDomain.end())` error.

I tried to debug this in many way.

<img width="929" alt="image" src="https://github.com/ANLAB-KAIST/KENSv3/assets/48939845/59867112-293d-4f30-819b-ce7699a58dd6">

First, I ran asan, refering to similar discussion, but it didn't make any error. It just failed assertion as same as test without asan.

Then, I tried to use debugger, and found strange thing.

<img width="476" alt="image" src="https://github.com/ANLAB-KAIST/KENSv3/assets/48939845/512fb543-4ecf-488e-8a21-fa3ec2e09a8b">

When I call createFileDescriptor, it returns fd=3.

<img width="453" alt="image" src="https://github.com/ANLAB-KAIST/KENSv3/assets/48939845/e2bc8508-59ea-47ec-90a9-2a320127fc91">

But, at the assertion failure point, I believe debugger says there's only fd=4. --I still keeped breakpoint created before, so no other createFileDescriptor has been called.--

I want to ask two questions :

1. Is there any other debugging method I can use to solve this problem? Or can you provide any hint about this?
2. If I simply remove removeFileDescriptor, I get full point at least on given random seed. Is it okay to not calling removeFileDescriptor if I can't find reason until deadline?

## Edit

Sorry for confusion. fd=4 was also created. But fd=3 should remained too.

**Comment by Pusnow** [comment]
> Is there any other debugging method I can use to solve this problem? Or can you provide any hint about this?

Insert `printf` in `createFileDescriptor` and `removeFileDescriptor` and trace all fd creation and deletion. If you really don't know why the problem persist, send your source code to TAs. cs341.pj@kaist.ac.kr


> If I simply remove removeFileDescriptor, I get full point at least on given random seed. Is it okay to not calling removeFileDescriptor if I can't find reason until deadline?

As long as you pass all test cases (including `test-kens-open-unreliable`), it is okay.
